#16037 Templated requirements.txt in Python operators

rounakdatta opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

Closes: https://github.com/apache/airflow/issues/16037
This pull request addresses the changes discussed in the aforementioned issue.

- [x] Added new unit tests
- [x] This change is not breaking
- [x] Updating documentation of the operator

Changes:
- Support for passing a `requirements.txt` file in the `requirements` field of the PythonVirtualenvOperator.
- The file can be named `*.txt` (anything ending with .txt), and can be jinja templated.

Note: `template_searchpath` must be appropriately set while using arbitrary locations of the template file.

rounakdatta committed:
Rebase to latest master

rounakdatta committed:
Address static check failures

rounakdatta committed:
Update param docs for the function

uranusjr commented on airflow/operators/python.py:
```diff
@@ -333,20 +337,31 @@ def __init__(
+            return
+
+        import pkg_resources
+        self.requirements = [str(req) for req in pkg_resources.parse_requirements(self.requirements)]
```
The requirements file format can do more than listing requirements, and those cannot be parsed by `parse_requirements` (I don’t remember whether setuptools would throw or just silently drop them.) So instead of trying to pre-parse the content, I would try to parse it to pip instead.

* Remove `pre_execute`
* Change `prepare_virtualenv` to accept `requirements` as str (plus the currently supported list form)
* When doing `pip install`, write the str type `requirements` to a temporary file for `pip intall -r`.

rounakdatta replied:
Ack, makes sense, let me work on that change.

rounakdatta replied:
@uranusjr I have tried an implementation of your proposed approach, can you once re-review?

rounakdatta committed:
Address review comments: use requirements.txt directly

rounakdatta committed:
Clean up: use only file interface for pip install

potiuk commented on airflow/utils/python_virtualenv.py:
```diff
@@ -87,14 +87,15 @@ def prepare_virtualenv(
     :type system_site_packages: bool
-    :param requirements: List of additional python packages
-    :type requirements: List[str]
+    :param requirements: Path to the requirements.txt file
```
I think we should still (backwards compatibility) handle the case where requirements are List[str]. It will make it a bit more complex, but I think it is needed. 

I propose to keep the old `requrements` handling only List[str] and add a new parameter `requirements_file_path`  - and check if only one of those is passed (and act accordingly). I think it is very ambiguous to name `requirements` something that is path to requirements file.

rounakdatta replied:
Addressed! 😃

potiuk reviewed (changes requested):
I think we need to support "prepare_virtualenv" backwards compatible behaviour.

rounakdatta committed:
Address review comments

Keep separate list & path string interfaces for passing virtualenv requirements

rounakdatta committed:
Fix a unit test

rounakdatta committed:
Fix: Prevent pip install upon empty

rounakdatta committed:
Fix test_add_dill UTs

turbaszek commented on airflow/operators/python.py:
```diff
@@ -300,7 +304,7 @@ def __init__(
         *,
         python_callable: Callable,
-        requirements: Optional[Iterable[str]] = None,
+        requirements: Union[Iterable[str], str, None] = None,
```
I think we usually tend to use `Optional` instead of `Union[..., None]` additionaly `Optional` marks the argument as... optional 😉 

```suggestion
        requirements: Optional[Union[Iterable[str], str]] = None,
```

rounakdatta replied:
Makes sense, addressed!

uranusjr replied:
FWIW the two are equivalent, `Optional[X]` is just syntax sugar to `Union[X, None]` (and IIRC Mypy normalises the former to latter internally, not sure)

turbaszek reviewed (approved):
Small comment but otherwise looks good 🚀 @rounakdatta would you mind rebasing?

Commit:
Merge branch 'apache:main' into 16037-support-requirements-txt

rounakdatta committed:
Address comment: Use Optional typehint

uranusjr commented on airflow/operators/python.py:
```diff
@@ -334,7 +338,7 @@ def __init__(
         *,
         python_callable: Callable,
-        requirements: Optional[Iterable[str]] = None,
+        requirements: Optional[Iterable[str], str] = None,
```
```suggestion
        requirements: Union[None, Iterable[str], str] = None,
```

uranusjr committed:
Fix Optional syntax

Commit:
Merge branch 'apache:main' into 16037-support-requirements-txt

uranusjr committed:
Merge branch 'main' into 16037-support-requirements-txt

uranusjr reviewed (approved):
Merging main to run the improved CI.

uranusjr committed:
Merge branch 'main' into 16037-support-requirements-txt

kaxil commented on airflow/operators/python.py:
```diff
@@ -405,6 +409,19 @@ def execute(self, context: Context):
         with TemporaryDirectory(prefix='venv') as tmp_dir:
+            requirements_file_name = f'{tmp_dir}/requirements.txt'
+
+            if isinstance(self.requirements, List):
```
```suggestion
            if isinstance(self.requirements, list):
```

doesn't make a difference but just for consistency across code-base. No strong preference

kaxil commented on airflow/utils/python_virtualenv.py:
```diff
@@ -89,12 +95,23 @@ def prepare_virtualenv(
     :type system_site_packages: bool
     :param requirements: List of additional python packages
     :type requirements: List[str]
+    :param requirements_file_path: Path to the requirements.txt file
```
```suggestion
    :param requirements_file_path: Path to the ``requirements.txt`` file
```

uranusjr committed:
Remove lazy-object-proxy reference in test_add_dill()

uranusjr committed:
Merge branch 'main' into 16037-support-requirements-txt

uranusjr committed:
Don't use pkg_resources in tests

This module is a part of setuptools, and not guaranteed to be installed
into a new virtual environment.

Fortunately funcsigs exposes the __version__ attribute at top level, so
we don't really need to read package metadata to test it.

uranusjr committed:
Docstring format

uranusjr committed:
Coerce 'requirements' in PythonVirtualenvOperator

uranusjr reviewed (approved):
I committed and pushed the previous review suggestions directly to the branch.

uranusjr committed:
Black
